src/Elliptic/DiscontinuousGalerkin/NumericalFluxes/FirstOrderInternalPenalty.hpp

```diff
@@ -0,0 +1,379 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <string>
+
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/Tensor/Metafunctions.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Domain/FaceNormal.hpp"
+#include "Domain/Tags.hpp"
+#include "NumericalAlgorithms/DiscontinuousGalerkin/NormalDotFlux.hpp"
+#include "NumericalAlgorithms/LinearOperators/Divergence.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+class DataVector;
+template <size_t>
+class Mesh;
+namespace Tags {
+template <typename>
+struct NormalDotFlux;
+template <typename>
+struct Normalized;
+}  // namespace Tags
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace elliptic {
+namespace dg {
+namespace NumericalFluxes {
+
+/*!
+ * \brief The internal penalty flux for first-order elliptic equations.
+ *
+ * \details Computes the internal penalty numerical flux (see e.g.
+ * \cite HesthavenWarburton, section 7.2) dotted with the interface unit normal.
+ *
+ * We implement here a suggested generalization of the internal penalty flux
+ * for any set of elliptic PDEs. It is designed for fluxes (i.e. principal parts
+ * of the PDEs) that may depend on the dynamic fields. This is the case for the
+ * velocity potential equation of binary neutron stars, for example.
+ *
+ * Formulating the elliptic PDEs in terms of a first-order _auxiliary_ variable
+ * \f$v\f$ for each _primal_ variable \f$u\f$, they take the _flux-form_
+ *
+ * \f{align}
+ * -\partial_i F^i_A + S_A = f_A
+ * \f}
+ *
+ * where the fluxes and sources are indexed by the variables and we have
+ * defined \f$f_A\f$ as those sources that are independent of the variables.
+ * We now choose the internal penalty numerical fluxes \f$F^{*i}\f$ as follows
+ * for each variable \f$u\f$ and their corresponding auxiliary variable \f$v\f$:
+ *
+ * \f{align}
+ * n_i F^{*i}_u &= \frac{1}{2} n_i \left( F^i_u(\partial_j
+ * F^j_v(u^\mathrm{int})) + F^i_u(\partial_j F^j_v(u^\mathrm{ext}))
+ * \right) - \sigma n_i \left(F^i_u(n\otimes u^\mathrm{int}) - F^i_u(
+ * n\otimes u^\mathrm{ext})\right) \\
```

Comment:
why tensor products instead of indices?

Reply:
Because one could mistake it for a free index.

Reply:
I find this notation to be opaque.  Do you have notes with the XCTS fluxes and sources defined for the first-order form?

Reply:
yes, this is a screenshot from the docs on another branch:

![Bildschirmfoto 2019-10-22 um 13 30 26](https://user-images.githubusercontent.com/746230/67281738-66685800-f4d0-11e9-9f60-a10022645086.png)

Does that help?

Reply:
I'm still confused about the penalty term. Can you please write Eq. (2) using only indices, not tensor products? Is F^i_u ( n x u^int) a product of F^i_u and (n x u^int) or saying F^i_u is a function of n x u^int?

Reply:
- Done, and included the missing F^i_v in the penalty term that you found in the other comment.
- The fluxes are functions of the primal and auxiliary variables, I added more detailed explanations to the docs.